node/db/src/rolling/gc.rs

```diff
@@ -0,0 +1,230 @@
+//! Time cost of `forest-cli snapshot export --dry-run` on DO droplet with 16GiB
+//! ram is between `9000s` to `11000s` for both scenarios, no significant
+//! performance regression has been observed
+//!
+//! ### Write performance
+//! DB write performance is typically on-par with `snapshot import`, note that
+//! when the `current` DB space is very large, it tends to trigger DB re-index
+//! more frequently, each DB re-index could pause the GC process for a few
+//! minutes. The same behaviour is observed during snapshot import as well.
+
+use std::{
+    sync::atomic::{self, AtomicU64, AtomicUsize},
+    time::Duration,
+};
+
+use chrono::Utc;
+use forest_blocks::Tipset;
+use forest_ipld::util::*;
+use forest_utils::db::BlockstoreBufferedWriteExt;
+use fvm_ipld_blockstore::Blockstore;
+use human_repr::HumanCount;
+use tokio::sync::Mutex;
+
+use super::*;
+
+pub struct DbGarbageCollector<F>
+where
+    F: Fn() -> Tipset + Send + Sync + 'static,
+{
+    db: RollingDB,
+    get_tipset: F,
+    lock: Mutex<()>,
+    gc_tx: flume::Sender<flume::Sender<anyhow::Result<()>>>,
+    gc_rx: flume::Receiver<flume::Sender<anyhow::Result<()>>>,
+    last_reachable_bytes: AtomicU64,
+}
+
+impl<F> DbGarbageCollector<F>
+where
+    F: Fn() -> Tipset + Send + Sync + 'static,
+{
+    pub fn new(db: RollingDB, get_tipset: F) -> Self {
+        let (gc_tx, gc_rx) = flume::unbounded();
+
+        Self {
+            db,
+            get_tipset,
+            lock: Default::default(),
+            gc_tx,
+            gc_rx,
+            last_reachable_bytes: AtomicU64::new(0),
+        }
+    }
+
+    pub fn get_tx(&self) -> flume::Sender<flume::Sender<anyhow::Result<()>>> {
+        self.gc_tx.clone()
+    }
+
+    /// This loop automatically triggers `collect_once` when the total DB size
+    /// is greater than 2x of the last reachable data size
+    pub async fn collect_loop_passive(&self) -> anyhow::Result<()> {
+        loop {
+            // Check every 10 mins
+            tokio::time::sleep(Duration::from_secs(10 * 60)).await;
+
+            // Bypass size checking during import
+            let tipset = (self.get_tipset)();
+            if tipset.epoch() == 0 {
+                continue;
+            }
+
+            // Bypass size checking when lock is held
+            {
+                let lock = self.lock.try_lock();
+                if lock.is_err() {
+                    continue;
+                }
+            }
+
+            if let (Ok(total_size), Ok(current_size), last_reachable_bytes) = (
+                self.db.total_size_in_bytes(),
+                self.db.current_size_in_bytes(),
+                self.last_reachable_bytes.load(atomic::Ordering::Relaxed),
+            ) {
+                let should_collect = if last_reachable_bytes > 0 {
+                    total_size > 2 * last_reachable_bytes
+                } else {
+                    total_size > 0 && current_size * 3 > total_size
```

Comment:
When we import a new snapshot, `last_reachable_bytes` will be 0 and `current_size` will be the same as `total_size`, right? Doesn't that mean that the GC will never be triggered?

Reply:
Oh, right, so it'll start a GC run immediately?

Reply:
The check interval is 10min, so it will be triggered 0-10min after snapshot being imported and HEAD being set